pkg/kubernetes/client/manifest.go

```diff
@@ -0,0 +1,112 @@
+package client
+
+import (
+	"bytes"
+	"errors"
+
+	yaml "gopkg.in/yaml.v2"
+)
+
+// Manifest represents a Kubernetes API object. The fields `apiVersion` and
+// `kind` are required, `metadata.name` should be present as well
+type Manifest map[string]interface{}
```

Comment:
This struct is a bit sad, what with the validation everywhere. Go makes it really difficult to avoid doing this, given you can't specify mandatory constructors, but one thing I've done before is a pattern like this:

https://gist.github.com/lawrencejones/cfbd66c47354ac5badcd2c100b46a5e6

Where you have a package for your `Manifest` struct, but it just wraps a package-private struct on a private field. In an ideal language, you'd be able to enforce that any variable with a `Manifest` type has all the required fields. This isn't possible in Go's type system, but by using this pattern you make it really difficult for people to avoid being aware that they're doing something wrong when creating a manifest object.

If I see there's a `Manifest` type, and want to create it, I'll first check how I can assign the fields:

```golang
manifest.Manifest{manifest: manifest.manifest{...}} // not going to work, manifest is a private field
```

When I grep through the code to find out how I should use it, the only thing associated with the `manifest.Manifest` (capitalised) type is a constructor, where we document that this is the only way you should ever construct these types.

It's not what you'd like, but Go doesn't make that possible. It does make it safe enough that you can omit the error checking in all the private methods though, as it's so unlikely someone is going to go out of their way to construct an invalid `Manifest` type.

Other alternatives are:
- Have `NewManifest() (*manifest, error)` return a private type. This gets pretty nasty though, as people need to define a lot of interfaces at the callsite
- Create a private struct field called `valid bool` in the manifest, and the only place `valid` is set to `true` is in the constructor that does the validation. Then wrap every struct method with:

```golang
func (m manifest) Namespace() string {
  if !m.valid {
    return ""
  }

  return m["metadata"].(map[string]interface{})["namespace"].(string)
}
```

This makes things slightly easier to read, given each method has the exact same validation step, so you can easily ignore the code.

I'm not sure if any of this is worth doing (I'd personally be tempted to just assume the struct is valid, and be careful of how I use it in the code) but wanted to share some options in case you were searching for some alternatives.

Reply:
Huh, we can't really use your idea because we rely heavily on `Manifest` being of type `map[string]interface{}` .. it MUST stay castable. 

I'm not completely happy with the current API though, it would be cool if we could assume that once we have a `Manifest` it's actually valid ... but what is even valid? `List` for example doesn't have a `metadata.name`. 

Probably this could be solved by implementing a custom `UnmarshalYAML`, because that's the only way to create a `Manifest` after all anyways.